pandas/core/arrays/datetimes.py

```diff
@@ -1953,7 +1953,13 @@ def sequence_to_dt64ns(
             data, inferred_tz = objects_to_datetime64ns(
                 data, dayfirst=dayfirst, yearfirst=yearfirst
             )
-            tz = _maybe_infer_tz(tz, inferred_tz)
+            try:
+                tz = _maybe_infer_tz(tz, inferred_tz)
+            except TypeError:
+                #  two timezones: convert to intended from base UTC repr
```

Comment:
i am not sure we actually want to do this, what is the current TypeError that you are catching here? we want to *very* explicity convert things

Reply:
TypeError is raised only the case that both timezones are present but do not match.

    raise TypeError(f"data is already tz-aware {inferred_tz}, unable to set specified tz: {tz}")

There is no reason, to my knowledge, why the information about the timezones and the underlying data cannot therefore be utilised to yield the correct dtype result, as in the tests. ?

Reply:
we are very explicit about these errors and -1
about catching at this level 

if it's wanted at a higher level i guess that's ok but again we need to be very explicit

Reply:
IIRC elsewhere we require the user to do `arr.tz_convert` rather than do this in the DTA/DTI constructor.  i dont have a particularly strong opinion on this

Reply:
This doesn't seem strictly true since:

    df = pd.DataFrame(data={"timezones": pd.date_range("2016-01-01", freq="d", periods=4, tz="US/Central")})

    df.astype({'timezones': 'datetime64[ns, Europe/Berlin]'})  # OK: performs the conversion
    df.astype({'timezones': 'datetime64[ns]'}) # OK: performs conversion to native UTC
    df.astype({timezones': 'object').astype({'timezones': 'US/Central'})  # OK: recovers the original from object str
    df.astype({timezones': 'object').astype({'timezones': 'Europe/Berlin'})  # FAILS: raises the TypeError
    df.astype({'timezones': 'object'}).astype({'timezones': 'datetime64[ns, US/Central]'}).astype({'timezones': 'datetime64[ns, Europe/Berlin]'})  # OK: performs the correct conversion

The change I have implemented basically just performs the fifth line instead of the fourth to recover the intended value.
Not sure why the TypeError is raised or its purpose when it is so easily bypassed and dealt with?

The method `_maybe_infer_tz` is only used twice in pandas, so I don't think this TypeError is useful, I think it just limits useful functionality.

Reply:
Re-written to exclude the Exception and simplify.

Reply:
> This doesn't seem strictly true since:

The examples you gave are all for `DataFrame.astype`, which has some unfortunate inconsistencies (xref #33401).  My comment is about what we allow in the constructor.

Reply:
@jreback I think I'm +1 on this because the behavior in the new test_astype_tz_object_conversion is something we should support

@attack68 this needs to be tested directly in tests.arrays.test_datetimes.TestSequenceToDT64NS and probably in TestDatetimeArrayConstructor